internal/messagejournal/messagejournal.go

```diff
@@ -0,0 +1,231 @@
+	const insertEntryTemplate string = `INSERT INTO %s (
+		message_id, sent, worker_name, response_to, worker_event, worker_message) 
+		values (?,?,?,?,?,?)`
+
+	insertAction, err := j.database.Prepare(fmt.Sprintf(insertEntryTemplate, messageJournalTableName))
+	if err != nil {
+		return nil, fmt.Errorf("cannot prepare statement for table '%v': %w", messageJournalTableName, err)
+	}
+
+	persistentResult, err := insertAction.Exec(
+		entry.MessageID,
+		entry.Sent,
+		entry.WorkerName,
+		entry.ResponseTo,
+		entry.WorkerEvent.EventName,
+		entry.WorkerEvent.EventMessage,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not insert journal entry into table '%v': %w", messageJournalTableName, err)
+	}
+
+	entryID, err := persistentResult.LastInsertId()
+	if err != nil {
+		return nil, fmt.Errorf("could not select last insert ID '%v' for table '%v': %w", entryID, messageJournalTableName, err)
+	}
+	log.Debugf("new message journal entry (id: %v) added: '%v'", entryID, entry.MessageID)
+
+	return &entry, nil
+}
+
+// GetEntries retrieves a list of all the journal entries in the message journal database
+// that meet the criteria of the provided message journal filter.
+func (j *MessageJournal) GetEntries(filter Filter) ([]map[string]string, error) {
+	entries := []map[string]string{}
+	queryString, err := j.buildDynamicGetEntriesQuery(filter)
+	if err != nil {
+		return nil, fmt.Errorf("cannot build dynamic sql query: %w", err)
+	}
+
+	preparedQuery, err := j.database.Prepare(queryString)
+	if err != nil {
+		return nil, fmt.Errorf("cannot prepare query when retrieving journal entries: %w", err)
+	}
+
+	rows, err := preparedQuery.Query()
+	if err != nil {
+		return nil, fmt.Errorf("cannot execute query to retrieve journal entries: %w", err)
+	}
+
+	for rows.Next() {
+		var rowID int
+		var messageID string
+		var sent time.Time
+		var workerName string
+		var responseTo string
+		var workerEvent uint
+		var workerEventMessage string
+
+		err := rows.Scan(
+			&rowID,
+			&messageID,
+			&sent,
+			&workerName,
+			&responseTo,
+			&workerEvent,
+			&workerEventMessage,
+		)
+		if err != nil {
+			return nil, fmt.Errorf("cannot scan journal entry columns: %w", err)
+		}
+
+		// Truncate the worker messages by the truncate length specified.
+		messageMaxSize := len(workerEventMessage)
+		if messageMaxSize >= filter.TruncateLength {
+			messageMaxSize = filter.TruncateLength
+			workerEventMessage = fmt.Sprintf("%+v...", workerEventMessage[:messageMaxSize])
+		}
+
+		// Convert the entry properties into a string format and append to the list of entries.
+		newMessage := map[string]string{
+			"message_id":     messageID,
+			"sent":           sent.String(),
+			"worker_name":    workerName,
+			"response_to":    responseTo,
+			"worker_event":   ipc.WorkerEventName(workerEvent).String(),
+			"worker_message": workerEventMessage,
+		}
+		entries = append(entries, newMessage)
```

Comment:
It could be very slow, when there will be thousands of entries. Have you tested this case?

Reply:
Yes it could be slow down with thousands of entries, but I was hoping the addition of filters would prevent a case where thousands of events would have to be retrieved at a single time (specifically the `message-id` or `to/sent` filters).

I'm not sure, is there an alternative way to handle this that you think would work well? :thinking:

Reply:
I was thinking about this. I wonder if there's a way to defer some of the
aggregation of this function further up the call stack. Maybe instead of
appending each entry to a slice within `GetEntries` and returning it at the end,
`GetEntries` could accept a callback function that gets called on each entry as
its read from the database. That passes the responsibility of aggregating the
results (or not) to the caller. I don't know what advantage that really brings
though, since eventually, this map gets serialized into a D-Bus message and
passed back across the D-Bus broker.